core/metadata.h

```diff
@@ -102,10 +50,114 @@ struct mt_attr {
   int scope_id;
 };
 
-void ComputeMetadataOffsets();
+typedef std::tuple<std::string, int> attr_id_t;
+
+class ScopeComponent {
+ public:
+  ScopeComponent()
+      : attr_id_(),
+        size_(),
+        offset_(),
+        scope_id_(),
+        assigned_(),
+        invalid_(),
+        modules_(),
+        degree_() {}
+
+  ~ScopeComponent(){};
+
+  const attr_id_t &attr_id() const { return attr_id_; };
+  void set_attr_id(attr_id_t id) { attr_id_ = id; };
+
+  int size() const { return size_; };
+  void set_size(int size) { size_ = size; };
+
+  mt_offset_t offset() const { return offset_; };
+  void set_offset(mt_offset_t offset) { offset_ = offset; };
+
+  scope_id_t scope_id() const { return scope_id_; };
+  void set_scope_id(scope_id_t id) { scope_id_ = id; };
+
+  bool assigned() const { return assigned_; };
+  void set_assigned(bool assigned) { assigned_ = assigned; };
+
+  bool invalid() const { return invalid_; };
+  void set_invalid(bool invalid) { invalid_ = invalid; };
+
+  const std::set<Module *> &modules() const { return modules_; };
+  void add_module(Module *m) { modules_.insert(m); };
+  void clear_modules() { modules_.clear(); };
+
+  int degree() const { return degree_; };
+  void incr_degree() { degree_++; };
+
+  bool DisjointFrom(const ScopeComponent &rhs);
+
+ private:
+  /* identification fields */
+  attr_id_t attr_id_;
+  int size_;
+  mt_offset_t offset_;
+  scope_id_t scope_id_;
+
+  /* computation state fields */
+  bool assigned_;
+  bool invalid_;
+  std::set<Module *> modules_;
+  int degree_;
+};
 
-char *GetScopeAttrName(scope_id_t scope_id);
-int *GetScopeAttrSize(scope_id_t scope_id);
+class Pipeline {
+ public:
+  Pipeline() : scope_components_(), module_scopes_(){};
+
+  ~Pipeline(){};
```

Comment:
Should be virtual. Perhaps we can remove this line and just rely on the default destructor, which is virtual?